storey/flow.py

```diff
@@ -320,21 +362,79 @@ def _user_fn_output_to_event(self, event, fn_result):
                 mapped_event.body = fn_result
             return mapped_event
 
-    def _check_step_in_flow(self, type_to_check):
+    def _check_step_in_flow(self, type_to_check, visited=None):
+        # initialize the visited set once at the top
+        if visited is None:
+            visited = set()
+
+        # detect cycles
+        if self in visited:
+            return False
+        visited.add(self)
+
+        # check this node
         if isinstance(self, type_to_check):
             return True
+
+        # check outlets
         for outlet in self._outlets:
-            if outlet._check_step_in_flow(type_to_check):
+            if outlet._check_step_in_flow(type_to_check, visited):
                 return True
+
+        # check recovery step
         if isinstance(self._recovery_step, Flow):
-            if self._recovery_step._check_step_in_flow(type_to_check):
+            if self._recovery_step._check_step_in_flow(type_to_check, visited):
                 return True
+
         elif isinstance(self._recovery_step, dict):
             for step in self._recovery_step.values():
-                if step._check_step_in_flow(type_to_check):
+                if step._check_step_in_flow(type_to_check, visited):
                     return True
         return False
 
+    def check_and_update_iteration_number(self, event) -> Optional[Callable]:
+        if hasattr(event, "_cyclic_counter") and self._max_iteration is not None:
+            counter = self.get_iteration_counter(event)
+            if counter >= self._max_iteration:
+                raise RuntimeError(f"Max iterations exceeded in step '{self.name}' for event {event.id}")
+            event._cyclic_counter[self.name] = counter + 1
+        else:
+            event._cyclic_counter = {self.name: 1}
+
+    def get_iteration_counter(self, event):
+        if not hasattr(event, "_cyclic_counter"):
+            return 0
+        else:
+            return event._cyclic_counter.get(self.name, 0)
```

Comment:
A more idiomatic way of doing this would be:
```suggestion
        counter = getattr(event, "_cyclic_counter", {}).get(self.name, 0)
```
which I think might as well be inlined.